Make ob-jq aware of org-tables (input and output)

Fuco1 opened the pull request:
A table formated as such

foo | bar
---|---
a | b
c | d

can serve as an input json which is an array with `foo` and `bar` being keys and `a,b` and `c,d` respective values.

The keyword for this is `:table <reference>`

Fuco1 commented:
We could probably make the `:stdin` option smart enough to detect if the reference is a block with text or a table... that would be another option to do this but maybe less explicit and more magical (although it has a certain coolness to it...).  What do you think about such an option?

Also the output code seems a bit peculiar to me so I've changed it a bit, was there any reason behind the  previous implementation?

Fuco1 commented:
I've also noticed I've introduced a dependency on `json-mode` for the result pretty-print... I think jq can do that by default so I'll try to fix that too.

ljos commented on ob-jq.el:
```diff
@@ -52,49 +53,66 @@
+           (with-temp-buffer
+             (insert results)
+             (json-mode)
+             (json-mode-beautify)
```
As you said, I also think jq can do this. I do use `json-mode` in `jq-interactively` to fontify the buffer, but only if the user has installed `json-mode`. I don't make it mandatory and I don't think we should when we have other options.

ljos commented:
As I understand it, there is good precedent for making the input a bit magic in org-babel; I think the smart `:stdin` is a good addition.

The whitespace fix is fine, but would you mind merging the other commits after fixing the hard dependency on `json-mode`?

Fuco1 commented:
Agreed on the magic, it always works in a "do what I want" manner so I think we should keep at it.

By merge you mean squashing the commits?  Sure, I'll do that, I accidentaly introduced some dependencies on my local config anyway :D

ljos commented:
Yes. I mean squashing them together. 
Thank you for working on this.

Fuco1 commented:
@ljos I've squashed the two commits adding the input support.

I've also extended the output so that with `:results table` it will try to reasonably guess if the output json is interpretable as a table and do that.

If it is a vector of (homogenous, assumed) objects, grab the keys as column names and generate a table.  If it is a vector of vectors just unpack the data into a table.  Otherwise print the result even if `:results table` is set as we can't guess what to do.

I've set the default to `:results output` so that without any special directives it always prints the output as recieved from `jq` (which also pretty prints it)

ljos commented on ob-jq.el:
```diff
@@ -49,52 +50,85 @@
+                        (insert
+                         (cond
+                          ((stringp res) res)
+                          ((listp res) (org-babel-jq-table-to-json res))))
```
I couldn't figure it out from the `org-babe-ref-resolve`; do we know that these two cases in the `cond` are the only cases that can happen? Could it return a number or other datatype? I don't mind not guarding for all of them, but it would help in debugging to have a better error there if someone comes asking. Right now if that happens it will give a very unfriendly error.

Fuco1 replied:
Agreed, I think it could return a number.  I think we can handle just a list or anything else which will just get passed along for org to do whatever it wants.  Good catch.

ljos replied:
Yes. When I look at the previous source that is what we do. Just add (t res) as a clause in the cond and we should be good.

ljos commented on ob-jq.el:
```diff
@@ -49,52 +50,85 @@
+                   (mapcar (lambda (row) (append row nil)) data))
+                  ;; If the first element is a list we will assume we
+                  ;; have an array of objects, so generate the colnames
+                  ;; accordingly
```
Is this an OK assumption? What happens if we get a list of different objects? F.ex. `[{"a" : 1}, {"b" : 2 }]`. 

I'm sorry to give you more work. I'm just not very familiar with the json-package and how it works so I can't immediately see what is going on here. If you don't have time I will check it myself later. I just need to be convinced it won't blow up in some horrible way when it gets some strange result data.

Fuco1 replied:
I think the `json` package is not relevant here.  The assumption is on the side of the user.  If we don't specify that we want a table result the output is always returned as "json string".  If the user specifies `:results table` then we assume the data is in a "table-compatible" way.

We can of course add an assertion and then default to some other type of output but maybe it's even better to error out and let the user know that "Hey, you want a table but we can't do that because the output is not a table, you probably have an error somewhere".

I can add a check that the returned vector is homogenous, that should be fairly simple.

ljos replied:
Yes, of course. I am sorry. I was quite tired yesterday. I was just worried that this would happen to all results and then it would be difficult to debug. I should have thought that it of course is only valid for tables. That is not a big problem. I'm not entirely sure we need a check, just a graceful error. I would assume the org function that writes the table will error for us if something is up.

ljos commented:
I am happy to merge this after you add the `(t res)` clause to the `cond` in `org-babel-execute:jq`.

Fuco1 commented:
@ljos I've fixed the cond and also added a catch-all clause for the table output: if it's not a vector of vectors or lists just pass the output (which is probably an array), such that

```org
#+NAME: foo
#+BEGIN_EXAMPLE
[1,2,3,4]
#+END_EXAMPLE

#+BEGIN_SRC jq :stdin foo :results table
.
#+END_SRC

#+RESULTS:
: [1 2 3 4]

```

still works at least somehow.  It could alternatively produce a table of one row.

Fuco1 commented:
Yea, I went ahead with that and just make it a table:

```org
#+NAME: foo
#+BEGIN_EXAMPLE
[1,2,3,4]
#+END_EXAMPLE

#+BEGIN_SRC jq :stdin foo :results table
.
#+END_SRC

#+RESULTS:
| 1 | 2 | 3 | 4 |
```

ljos commented:
Thank you very much for working on this! I added you as a collaborator so you can make changes in the future or fix bugs without asking if you need to. You don't have to accept if you don't feel like it or have other things you would rather spend time on.

Fuco1 commented:
Cool!  I'll probably run things through PRs anyway as it's good to have more eyes on the changes (except for the trivial fixes like typos etc).

ljos commented:
I think that is a good idea. Thanks for contributing!
